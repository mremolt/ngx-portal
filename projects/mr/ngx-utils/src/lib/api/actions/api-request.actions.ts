import { HttpErrorResponse, HttpRequest } from '@angular/common/http';
import { Action } from '@ngrx/store';

import { ApiActions, ApiRequestConfig } from '../types';

export enum ApiRequestActionTypes {
  ApiRequest = '[API] Request',
  ApiError = '[API] Error',
}

export class ApiRequest<T = any> implements Action {
  public readonly type = ApiRequestActionTypes.ApiRequest;

  constructor(public payload: ApiRequestConfig<T>) {}
}

export class ApiError implements Action {
  public readonly type = ApiRequestActionTypes.ApiError;

  constructor(public payload: HttpErrorResponse) {}
}

export class ApiGetAction<T> extends ApiRequest {
  constructor(url: string, actions: ApiActions<T>) {
    const request = new HttpRequest<T>('GET', url);
    super({ request, actions });
  }
}

export class ApiPostAction<T> extends ApiRequest {
  constructor(url: string, body: T | null, actions: ApiActions<T>) {
    const request = new HttpRequest<T>('POST', url, body);
    super({ request, actions });
  }
}

export class ApiPutAction<T> extends ApiRequest {
  constructor(url: string, body: T | null, actions: ApiActions<T>) {
    const request = new HttpRequest<T>('PUT', url, body);
    super({ request, actions });
  }
}

export class ApiDeleteAction<T> extends ApiRequest {
  constructor(url: string, actions: ApiActions<T>) {
    const request = new HttpRequest<T>('DELETE', url);
    super({ request, actions });
  }
}
