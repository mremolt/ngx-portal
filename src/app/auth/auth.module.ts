import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { EffectsModule } from '@ngrx/effects';
import { StoreModule } from '@ngrx/store';

import { AuthEffects } from './reducers/auth.effects';
import { AuthGuard } from './reducers/auth.guard';
import * as fromAuth from './reducers/auth.reducer';

@NgModule({
  imports: [
    CommonModule,
    StoreModule.forFeature('auth', fromAuth.reducer),
    EffectsModule.forFeature([AuthEffects]),
  ],
  declarations: [],
  providers: [AuthGuard],
})
export class AuthModule {}
