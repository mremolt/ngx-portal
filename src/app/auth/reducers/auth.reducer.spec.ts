import { Authenticate, AuthenticateError, AuthenticateSuccess, Logout } from './auth.actions';
import { initialState, reducer } from './auth.reducer';

describe('Auth Reducer', () => {
  describe('unknown action', () => {
    it('should return the initial state', () => {
      const action = {} as any;

      const result = reducer(initialState, action);

      expect(result).toBe(initialState);
    });

    it('should return the initial state on undefined state', () => {
      const action = {} as any;

      const result = reducer(undefined, action);

      expect(result).toBe(initialState);
    });
  });

  describe('Authenticate action', () => {
    it('should set loading to true', () => {
      const action = new Authenticate('test', 'pass');

      const result = reducer(initialState, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('AuthenticateSuccess action', () => {
    it('should set token', () => {
      const action = new AuthenticateSuccess({ accessToken: '12345' });

      const result = reducer(initialState, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('AuthenticateError action', () => {
    it('should set error', () => {
      const action = new AuthenticateError('payload' as any);

      const result = reducer(initialState, action);

      expect(result).toMatchSnapshot();
    });

    it('should empty token', () => {
      const action = new AuthenticateError('payload' as any);
      const currentState = { ...initialState, token: { accessToken: '12345' } };

      const result = reducer(currentState, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('Logout action', () => {
    it('should set initialState', () => {
      const action = new Logout();
      const currentState = { ...initialState, token: { accessToken: '12345' } };

      const result = reducer(currentState, action);

      expect(result).toBe(initialState);
    });
  });
});
