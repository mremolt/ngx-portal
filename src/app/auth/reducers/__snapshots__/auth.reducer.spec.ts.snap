// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Reducer Authenticate action should set loading to true 1`] = `
Object {
  "error": null,
  "loading": true,
  "token": Object {
    "accessToken": "",
  },
}
`;

exports[`Auth Reducer AuthenticateError action should empty token 1`] = `
Object {
  "error": "payload",
  "loading": false,
  "token": Object {
    "accessToken": "",
  },
}
`;

exports[`Auth Reducer AuthenticateError action should set error 1`] = `
Object {
  "error": "payload",
  "loading": false,
  "token": Object {
    "accessToken": "",
  },
}
`;

exports[`Auth Reducer AuthenticateSuccess action should set token 1`] = `
Object {
  "error": null,
  "loading": false,
  "token": Object {
    "accessToken": "12345",
  },
}
`;
